### PR TITLE
[Feature] add two fraud datasets

### PR DESCRIPTION
**Pull request by kayzliu**
## Description

Added two fraud datasets, Yelp and Amazon. #2876 
@zhjwy9343 

## Checklist
Please feel free to remove inapplicable items for your PR.
- [x] The PR title starts with [$CATEGORY] (such as [NN], [Model], [Doc], [Feature]])
- [x] Changes are complete (i.e. I finished coding on this PR)
- [x] All changes have test coverage
- [x] Code is well-documented
- [x] To the my best knowledge, examples are either not affected by this change,
      or have been fixed to be compatible with this change
- [x] Related issue is referred in this PR
- [ ] If the PR is for a new model/paper, I've updated the example index [here](../examples/README.md).

### CODE REVIEW

File: python/dgl/data/fraud.py [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+        test_mask = torch.zeros(N, dtype=torch.bool)
+        train_mask[train_idx] = True
+        val_mask[val_idx] = True
+        test_mask[test_idx] = True
+        self.train_mask = train_mask
+        self.val_mask = val_mask
+        self.test_mask = test_mask
+    
+class FraudYelpDataset(FraudDataset):
+    def __init__(self, raw_dir=None, random_seed=2, train_size=0.7, val_size=0.1):
```
Comment (zhjwy9343) [inline python/dgl/data/fraud.py:139]: Please refer to the other datasets' description, e.g. https://docs.dgl.ai/_modules/dgl/data/gnn_benckmark.html#AmazonCoBuyComputerDataset. Give some statistics of the data, context of the dataset, and examples of the dataset.

File: python/dgl/data/fraud.py [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+        super(FraudYelpDataset, self).__init__(name='yelp',
+                                           raw_dir=raw_dir,
+                                           random_seed=random_seed,
+                                           train_size=train_size,
+                                           val_size=val_size)
+                                           
+class FraudAmazonDataset(FraudDataset):
+    def __init__(self, raw_dir=None, random_seed=2, train_size=0.7, val_size=0.1):
```
Comment (zhjwy9343) [inline python/dgl/data/fraud.py:147]: ditto

**Review by zhjwy9343 — CHANGES_REQUESTED** [review]

File: python/dgl/data/fraud.py [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+        for relation in self.relations[self.name]:
+            g = from_scipy(data[relation])
+            graphs.append(g)
+        self.graphs = graphs
+        
+        self.feature = node_features
+        self.label = node_labels
+        self._random_split(self.feature, self.label, self.seed, self.train_size, self.val_size)
+
+    def __getitem__(self, idx):
```
Comment (classicsong) [inline python/dgl/data/fraud.py:81]: You need to explain what is stored in self.graphs. There are multiple graphs, each for one relation type.
Furthermore, it is better to give an example of how to use these graphs.

**Review by classicsong — COMMENTED** [review]
In general, I do not catch the point why you store dataset as a list of graphs other than a DGL heterogeneous graph.

File: python/dgl/data/fraud.py (hunk 3 above)
Comment (kayzliu) [reply python/dgl/data/fraud.py:115]: Comments on __getitem__ and examples have been added.

File: python/dgl/data/fraud.py (hunk 1 above)
Comment (kayzliu) [reply python/dgl/data/fraud.py:245]: Thank you for the review, the statistics, context, and examples of the dataset have been added in latest commit. If there are any further problems, please let me know.

File: python/dgl/data/fraud.py [hunk 4]
```diff
@@ -0,0 +1,152 @@
...
+import numpy as np
+
+from .utils import save_graphs, save_info, load_graphs, load_info, _get_dgl_url
+from ..convert import from_scipy
+from .dgl_dataset import DGLBuiltinDataset
+
+class FraudDataset(DGLBuiltinDataset):
+    """Fraud node prediction dataset.
+
+    The dataset includes two homogeneous multi-relational graphs extracted from
```
[the remark this answers is not in the document]
Comment (kayzliu) [reply python/dgl/data/fraud.py]: Thanks for your question. The graph is a multi-relational graph, which has only one type of nodes and multiple types of edges. I have change it to heterogeneous graph in latest commit.

File: python/dgl/data/fraud.py [hunk 5]
```diff
@@ -0,0 +1,152 @@
...
+    def __len__(self):
+        # number of data examples
+        return len(self.graphs)
+
+    def save(self):
+        # save processed data to directory `self.save_path`
+        graph_path = os.path.join(self.save_path, self.name + '_dgl_graph.bin')
+        info_path = os.path.join(self.save_path, self.name + '_dgl_graph.pkl')
+        save_graphs(str(graph_path), self.graphs)
+        save_info(info_path, {'label': self.label, 'feature': self.feature})
```
[the remark this answers is not in the document]
Comment (kayzliu) [reply python/dgl/data/fraud.py]: Modified based on your comments.

File: python/dgl/data/fraud.py [hunk 6]
```diff
@@ -0,0 +1,152 @@
...
+        Directory to save the processed dataset.
+        Default: the value of `raw_dir`
+    """
+    file_urls = {
+        'yelp' : 'dataset/FraudYelp.zip',
+        'amazon' : 'dataset/FraudAmazon.zip'
+    }
+    relations = {
+        'yelp' : ['net_rsr', 'net_rtr', 'net_rur'],
+        'amazon' : ['homo', 'net_upu', 'net_usu', 'net_uvu']
```
[the remark this answers is not in the document]
Comment (kayzliu) [reply python/dgl/data/fraud.py]: Added in the doc string.

File: python/dgl/data/fraud.py [hunk 7]
```diff
@@ -0,0 +1,152 @@
...
+
+    def process(self):
+        # process raw data to graphs, labels, splitting masks
+        file_path = os.path.join(self.raw_path, self.file_names[self.name])
+        
+        data = io.loadmat(file_path)
+        node_features = torch.from_numpy(data['features'].todense())
+        node_labels = torch.from_numpy(data['label'])
+        
+        N = node_labels.shape[0]
```
[the remark this answers is not in the document]
Comment (kayzliu) [reply python/dgl/data/fraud.py]: Deleted.

File: python/dgl/data/fraud.py [hunk 8]
```diff
@@ -0,0 +1,313 @@
...
+
+        data = io.loadmat(file_path)
+        node_features = torch.from_numpy(data['features'].todense())
+        node_labels = torch.from_numpy(data['label'])
+        node_labels = node_labels.transpose(0, 1)
+
+        graph_data = {}
+        for relation in self.relations[self.name]:
+            u, v, _, _ = graphdata2tensors(data[relation])
+            graph_data[('node', relation, 'node')] = (u, v)
```
[the remark this answers is not in the document]
Comment (kayzliu) [reply python/dgl/data/fraud.py]: As per our discussion, this component is modified.

File: python/dgl/data/fraud.py [hunk 9]
```diff
@@ -0,0 +1,313 @@
...
+
+    Examples
+    --------
+    >>> dataset = FraudDataset('yelp')
+    >>> graph = dataset[0]
+    >>> num_classes = dataset.num_classes
+    >>> feat = dataset.ndata['feature']
+    >>> label = dataset.ndata['label']
+    """
+    file_urls = {
```
[the remark this answers is not in the document]
Comment (kayzliu) [reply python/dgl/data/fraud.py:72]: As per our discussion, this component is not going to change.

File: python/dgl/data/fraud.py [hunk 10]
```diff
@@ -0,0 +1,313 @@
...
+        """
+        assert idx == 0, "This dataset has only one graph"
+        return self.graph
+
+    def __len__(self):
+        """number of data examples"""
+        return len(self.graph)
+
+    @property
+    def num_classes(self):
```
[the remark this answers is not in the document]
Comment (kayzliu) [reply python/dgl/data/fraud.py:147]: As per our discussion, this component is not going to change.

File: python/dgl/data/fraud.py [hunk 11]
```diff
@@ -0,0 +1,313 @@
...
+
+    def has_cache(self):
+        """check whether there are processed data in `self.save_path`"""
+        graph_path = os.path.join(self.save_path, self.name + '_dgl_graph.bin')
+        return os.path.exists(graph_path)
+
+    def _random_split(self, x, node_labels, seed=2, train_size=0.7, val_size=0.1):
+        """split the dataset into training set, validation set and testing set"""
+        N = x.shape[0]
+        if self.name == 'yelp':
```
[the remark this answers is not in the document]
Comment (kayzliu) [reply python/dgl/data/fraud.py]: As per our discussion, this component is modified.

File: python/dgl/data/fraud.py [hunk 12]
```diff
@@ -0,0 +1,317 @@
...
+
+        self._random_split(g.ndata['feature'], g.ndata['label'], self.seed, self.train_size, self.val_size)
+
+    def __getitem__(self, idx):
+        r""" Get graph object
+
+        Parameters
+        ----------
+        idx : int
+            Item index, KarateClubDataset has only one graph object
```
[the remark this answers is not in the document]
Comment (kayzliu) [reply python/dgl/data/fraud.py]: Deleted.

File: python/dgl/data/fraud.py [hunk 13]
```diff
@@ -0,0 +1,317 @@
...
+    def has_cache(self):
+        """check whether there are processed data in `self.save_path`"""
+        graph_path = os.path.join(self.save_path, self.name + '_dgl_graph.bin')
+        return os.path.exists(graph_path)
+
+    def _random_split(self, x, node_labels, seed=2, train_size=0.7, val_size=0.1):
+        """split the dataset into training set, validation set and testing set"""
+        N = x.shape[0]
+        index = list(range(N))
+        train_idx, test_idx, _, y = train_test_split(index, node_labels, stratify=node_labels,
```
[the remark this answers is not in the document]
Comment (kayzliu) [reply python/dgl/data/fraud.py]: Modified based on your comments.

File: python/dgl/data/fraud.py [hunk 14]
```diff
@@ -0,0 +1,317 @@
...
+                                                         train_size=train_size,
+                                                         random_state=seed, shuffle=True)
+        if self.name == 'amazon':
+            # 0-3304 are unlabeled nodes
+            index = list(range(3305, N))
+            train_idx, test_idx, _, y = train_test_split(index, node_labels[3305:], stratify=node_labels[3305:],
```
[the remark this answers is not in the document]
Comment (kayzliu) [reply python/dgl/data/fraud.py]: Modified based on your comments.

File: python/dgl/data/fraud.py [hunk 15]
```diff
@@ -0,0 +1,317 @@
...
+                                                         test_size=train_size, random_state=seed, shuffle=True)
+
+        val_idx, test_idx, _, _ = train_test_split(test_idx, y, stratify=y, train_size=val_size / (1 - train_size),
```
[the remark this answers is not in the document]
Comment (kayzliu) [reply python/dgl/data/fraud.py]: Modified based on your comments.

**Review by zhjwy9343 — APPROVED** [review]
LGTM

**Pull request merged**